Emoji Picker Menu Navigation

jasperhuangg opened the pull request:
### Details
Added the ability to navigate through the emoji picker menu with arrow keys, and to send the highlighted emoji with [Enter]. It should function the same as the emoji picker in Slack.
- the highlighted emoji should never be cutoff by the window
- the first emoji is highlighted upon searching emojis
- hovering over an emoji with your cursor will also change the currently highlighted emoji
- if no emoji is highlighted, pressing an arrow key will highlight the first emoji
- the cursor in the search input is not affected by arrow key presses that change the highlighted emoji

### Fixed Issues
Fixes https://github.com/Expensify/Expensify.cash/issues/2450

### Tests

I tested this feature on Web and Desktop since those are the only two places it can actually be used, but I also tested it on Mobile Web to ensure that the touchscreen check works and that the Emoji Picker Menu doesn't crash.

1. Opened the emoji picker. Scroll down till you can see the end of on category and the start of another.
2. Use your mouse to hover over an emoji near this boundary.
3. Use the up/down arrow keys to move the highlight around this border. Verified that it skips over the header.
4. Move the highlight at the top and bottom edges of the window with the arrow keys. Verified that the menu scrolls to keep the highlighted emoji in the window.
5. Type in search terms to filter by. Verified that the first search result is highlighted. Verified that nothing breaks when you try to go out of bounds. Verified that you can move the cursor using the appropriate arrow keys when they would move the highlight out of bounds.
6. Pressed [Enter]. Verified that the highlighted emoji was inputted into the compose box.
7. Repeat steps 1-6 two more times to verify that event listeners are removed and added correctly.

### QA Steps
Identical to the tests above. See videos for examples.

### Tested On

- [x] Web
- [ ] Mobile Web
- [x] Desktop
- [ ] iOS
- [ ] Android

### Screenshots

#### Web
https://user-images.githubusercontent.com/31285285/115820241-0048f200-a433-11eb-9dd8-d6c274158365.mp4

#### Desktop
https://user-images.githubusercontent.com/31285285/115819992-8ca6e500-a432-11eb-9322-4cabf9f107f3.mp4

cc @stitesExpensify

jasperhuangg committed:
Add ability to highlight emoji, move around emoji picker

jasperhuangg committed:
Fit highlighted emoji into scroll window

jasperhuangg committed:
Use React.memo

jasperhuangg committed:
use onScroll

jasperhuangg committed:
DRY

jasperhuangg committed:
reorder functions

jasperhuangg committed:
Add ability to press enter to send

jasperhuangg committed:
Add consts

jasperhuangg committed:
Rename const

jasperhuangg committed:
Change function def

jasperhuangg committed:
Remove unused

jasperhuangg committed:
Add comment for React.memo

jasperhuangg committed:
Fix check for keypresses

jasperhuangg committed:
Reorder

jasperhuangg committed:
hover to updated the highlighted index, add mode where no emoji is highlighted

jasperhuangg committed:
Prevent arrow key presses from moving the cursor when they're changing the highlighted emoji.

jasperhuangg committed:
style

jasperhuangg committed:
style

jasperhuangg committed:
renaming and comments

jasperhuangg committed:
Add cleanup for event listeners

pecanoro commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -61,6 +90,128 @@ class EmojiPickerMenu extends Component {
+        // If nothing is highlighted and an arrow key is pressed
+        // select the first emoji
+        if (this.state.highlightedIndex === -1) {
+            this.setState(prevState => ({highlightedIndex: prevState.filteredEmojis.length === emojis.length ? 8 : 0}));
```
Why 8? Sorry, I just don't know what it means. Maybe leaving a comment about it would be nice.

pecanoro replied:
Ohh, it's the name of columns, you should use `numColumns` instead of hardcoding the number

jasperhuangg replied:
Good catch! I completely missed this but this is definitely what I was trying to do.

pecanoro commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -61,6 +90,128 @@ class EmojiPickerMenu extends Component {
+            if (boundsCheck()) {
+                return;
+            }
+            const isHeader = e => e.header || e.code === CONST.EMOJI_SPACER;
+            do {
+                newIndex += steps;
+            } while (isHeader(this.state.filteredEmojis[newIndex]));
```
what is this doing? Not a blocker or anything, but a comment may make things easier to understand for non react native experts.

jasperhuangg committed:
Move setState logic for shouldDisablePointerEvents into a more fitting place, add comment

jasperhuangg committed:
Add comment

jasperhuangg committed:
Use this.numColumns

jasperhuangg committed:
Fix comment

jasperhuangg committed:
Fix comments

jasperhuangg committed:
Fix comments

jasperhuangg committed:
Update comments

jasperhuangg committed:
Update comments

jasperhuangg committed:
Update comments

jasperhuangg commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -61,6 +91,138 @@ class EmojiPickerMenu extends Component {
+     * Highlights emojis adjacent to the currently highlighted emoji depending on the arrowKey
+     * @param {String} arrowKey
+     */
+    highlightAdjacentEmoji(arrowKey) {
```
Looking back the name of this function is a bit confusing; its purpose is just to move the highlight to a new index depending on which arrow key is pressed. Do you guys have any recommendations on better names (if you think a name change is necessary).

@pecanoro @stitesExpensify

pecanoro replied:
Nah, I think the name is pretty good, I would leave it like this

jasperhuangg committed:
Update comments

jasperhuangg commented on src/pages/home/report/EmojiPickerMenuItem.js:
```diff
@@ -3,28 +3,46 @@ import PropTypes from 'prop-types';
+    onHover: PropTypes.func.isRequired,
+
+    // Whether this menu item is currently highlighted or not
+    isHighlighted: PropTypes.bool.isRequired,
```
Just for future reference, what are the different scenarios where I should be using `isHighlighted` vs. `shouldHighlight` for this and other similar boolean props? 

@pecanoro @stitesExpensify

pecanoro replied:
IMO I would use `isHighlighted` which it's a variable that represents the current state of an element and I would use `shouldHighlight` when it's something that we want to apply to a future element/variable, for example, passing it as a parameter to a function.

jasperhuangg replied:
Thanks!

jasperhuangg commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -61,6 +91,138 @@ class EmojiPickerMenu extends Component {
+        // Setup and attach keypress/mouse handlers only if we have a keyboard (and not a touchscreen)
+        // NOTE: these event handlers are instance members so we can reference
+        // the same functions in memory when removing them.
+        if (!canUseTouchScreen() && document) {
```
I tested this on Mobile Web, and it turns out that Mobile Web also uses `index.native.js` instead of this file. Since this file will only be referenced by Desktop and Web (which should never have touchscreens and should always have a `document` object available), do you think it's still necessary to have this check?

@pecanoro @stitesExpensify

stitesExpensify replied:
hmm how did you come to that conclusion? I'm 90% sure that mobile web uses index.js

jasperhuangg replied:
I tried it on Safari in the simulator, and the search input doesn't appear (implying we're using `index.native.js`).

jasperhuangg replied:
Ahh wait, when I tried it on Safari in the simulator there was an error causing the search input not to appear leading me to think we were using `index.native.js`. Looks like it's still there and I need to include this check 👍

stitesExpensify replied:
This is actually what prevents the search bar 😄 https://github.com/Expensify/Expensify.cash/pull/2532/files#diff-0aeb1e7eb3f2813b84c81ee9aaf16b4e67307297dbc129803f5f0eb20a6a3693R313

In index.js we have a `isSmallScreenWidth`

jasperhuangg committed:
Merge branch 'main' into jasper-emojiPickerArrowKeys

jasperhuangg commented:
@pecanoro @stitesExpensify Thanks for clarifying my questions! Should be good for another review.

pecanoro commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -44,12 +48,38 @@ class EmojiPickerMenu extends Component {
             headerIndices: this.unfilteredHeaderIndices,
+            highlightedIndex: -1,
+            currentScrollOffset: 0,
+            shouldDisablePointerEvents: false,
```
Why is this one called `shouldDisablePointerEvents` and not `arePointerEventsDisabled`

jasperhuangg replied:
Hmmm that's a good point, since this is updated by the callback that actually disables pointer events. I guess because the boolean is used in actually disabling the pointer events, but it makes more sense to call it that.

pecanoro commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -92,32 +279,37 @@ class EmojiPickerMenu extends Component {
     }
 
     render() {
+        const pointerEvents = this.state.shouldDisablePointerEvents ? 'none' : 'auto';
```
What are the pointerEvents?

jasperhuangg replied:
It's essentially a string passed into the props for a `View` directly underneath. It basically disables the ability for the View to receive pointer events (in this case hover events). You can think of this as the `pointerEvents` css property, which is either `none` or `auto` depending on whether you want them to be enabled or not. I can update the name to `pointerEventsStr` to clarify things if you think that's better!

pecanoro replied:
Oh you didn't have to change it, I simply didn't know what they were hehe Maybe a comment would have been enough, I am just trying to learn more how things in React Native works, all this UI stuff gets me a bit confused.

jasperhuangg committed:
merge master

jasperhuangg committed:
Remove unused pointerEvents variable, change name of pointer events disabled state

jasperhuangg committed:
Merge remote-tracking branch 'origin/jasper-emojiPickerArrowKeys' into jasper-emojiPickerArrowKeys

# Conflicts:
#	src/CONST.js

jasperhuangg committed:
Remove whitespace

stitesExpensify commented:
Tested and it works great! Just some code organization comments and then we should be good to go!

jasperhuangg commented on src/CONST.js:
```diff
@@ -114,6 +114,9 @@ const CONST = {
     },
 
     EMOJI_PICKER_SIZE: 360,
+    EMOJI_PICKER_LIST_HEIGHT: 300,
```
Good catch, that's definitely not clear 👍

jasperhuangg committed:
Change const name

jasperhuangg committed:
Get rid of touchscreen check

jasperhuangg commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -44,12 +48,38 @@ class EmojiPickerMenu extends Component {
+        this.keyToDefaultPreventer = {};
+        ['ArrowDown', 'ArrowLeft', 'ArrowRight', 'ArrowUp'].forEach((key) => {
+            this.keyToDefaultPreventer[key] = (keyBoardEvent) => {
+                if (keyBoardEvent.key === key) {
```
I've defined them in an object so that we can reference the same event handler for each arrow key. That way we can always be adding and removing the same event handlers in memory. We don't need to check if the key is in the array since `toggleKeysOnSearchInput` is only ever called with arrow keys.

stitesExpensify replied:
Doing it this way we're adding 4 event handlers to the `keyDown` event though so every time a key is pressed we run through the same method 4 times. I think a better way to do it would be to just have one method that is connected to `keyDown` and then keep the active keys in the state rather than connecting and disconnecting keyboard events. Something like this
```
onKeyPressed(key): {
    if (this.state.keysToAccept.contains(keyBoardEvent.key) {
        keyBoardEvent.preventDefault();
    }
}
```

And then instead of `this.toggleKeysOnSearchInput` as a method, you can just set the state to include the relevant keys [here](https://github.com/Expensify/Expensify.cash/pull/2532/files#diff-0aeb1e7eb3f2813b84c81ee9aaf16b4e67307297dbc129803f5f0eb20a6a3693R272-R276).

jasperhuangg replied:
Gootcha, thanks for clarifying what you meant!

jasperhuangg commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -44,12 +48,38 @@ class EmojiPickerMenu extends Component {
+                }
+            };
+        });
+        this.toggleKeysOnSearchInput = (keysToIgnore = [], keysToAccept = []) => {
```
I just figured it would be clearer for `toggleKeysOnSearchInput` to go next to `keyToDefaultPreventer` since they're directly related. We have to construct `keyToDefaultPreventer` inside of the constructor.

jasperhuangg commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -61,6 +91,138 @@ class EmojiPickerMenu extends Component {
+        // NOTE: these event handlers are instance members so we can reference
+        // the same functions in memory when removing them.
+        if (!canUseTouchScreen() && document) {
+            this.keyDownHandler = (e) => {
```
What do you mean? Are you saying this should be stuck inside of a class-scoped function that's called when the component mounts?

stitesExpensify replied:
Right, I think it should be a defined function instead of just being defined inside of `componentDidMount`

jasperhuangg committed:
Move event handler setup into function to cleanup componentDidMount

jasperhuangg commented:
Hey @stitesExpensify thanks so much for the review! I've addressed most of your concerns but have a few clarifications of my own. Let me know what you think!

stitesExpensify commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -61,6 +91,138 @@ class EmojiPickerMenu extends Component {
+        // Setup and attach keypress/mouse handlers only if we have a keyboard (and not a touchscreen)
+        // NOTE: these event handlers are instance members so we can reference
+        // the same functions in memory when removing them.
+        if (!canUseTouchScreen() && document) {
```
Hmm I see a commit that says "remove canUseTouchScreen" but it looks like it's still in a lot of places in this file. Did something go wrong in a following commit?

jasperhuangg replied:
Wooops, really not sure what happened but this should be fixed now.

jasperhuangg committed:
Remove canUseTouchScreen

jasperhuangg committed:
Cleanup prevent default behavior for arrow key presses on the search input.

jasperhuangg commented:
@stitesExpensify Ready for another review, thanks for your feedback!

stitesExpensify reviewed (approved):
Looks great! Thanks for making all of those changes

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/Expensify.cash/actions/runs/807521239) to staging in version: 1.0.35-2🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

isagoico commented:
## Emoji Picker - Moving by arrow keys will eventually lose selection

## Expected result
Focus isn't lost even after reaching the last emoji on the list. 

## Actual result 
After reaching the last emoji on the list, the focus is not visible anymore. 

## Action Performed
1. Log in to expensify.cash
2. Navigate to a conversation
3. Click on the emoji icon
4. Use the arrow keys to reach the bottom of the list

## Platform

Web ✔️ 

**Build**: 1.0.36-0

**Notes/Images/Video**

https://user-images.githubusercontent.com/44479856/116945606-15e5c500-ac46-11eb-9e65-ce28fcc1e469.mp4

isagoico commented:
@jasperhuangg Not sure if the issue above should be a deploy blocker, let me know if I should open as a separate issue.

roryabraham commented:
No need for this to be a deploy blocker, @isagoico, can you create a separate issue and tag @jasperhuangg?

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/Expensify.cash/actions/runs/822089210) to production in version: 1.0.39-5🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
